MultiLangProgGuide/MultiLangProgGuide.rst

```diff
@@ -1819,6 +1821,1189 @@ If the requested :code:`timeout_nsec` exceeds
+zero length string if the structure has not yet been allocated, i.e.,
+:code:`cbuft` is :code:`nil`), and a return code of :code:`C.YDB_OK`.
+
+Go BufferT SetValBAry()
+-----------------------
+
+.. code-block:: go
+
+    SetValBAry(val *[]byte) int
+
+
+If the length of :code:`val` is greater than the :code:`len_alloc`
+field of the :code:`C.ydb_buffer_t` structure referenced by
+:code:`cbuft`, make no changes and return :code:`C.YDB_ERR_INVSTRLEN`.
+
+Otherwise, copy the bytes of :code:`val` to the location referenced by
+the :code:`buf_addr` field of the :code:`C.ydbbuffer_t` structure, set
+the :code:`len_used` field to its length and return with a return code
+of :code:`C.YDB_OK`.
+
+Go BufferT SetValStr()
+----------------------
+
+.. code-block:: go
+
+    SetVarStr(val *string) int
+
+If the length of :code:`val` is greater than the :code:`len_alloc`
+field of the :code:`C.ydb_buffer_t` structure referenced by
+:code:`cbuft`, make no changes and return :code:`C.YDB_ERR_INVSTRLEN`.
+
+Otherwise, copy the bytes of :code:`val` to the location referenced by
+the :code:`buf_addr` field of the :code:`C.ydbbuffer_t` structure, set
+the :code:`len_used` field to its length and return with a return code
+of :code:`C.YDB_OK`.
+
+----------------------------------
+Go Access Methods for BufferTArray
+----------------------------------
+
+Go BufferTArray Alloc()
+-----------------------
+
+.. code-block:: go
+
+    Alloc(numSubs, bufSiz uint)
+
+Allocate:
+
+- :code:`numSubs` buffers in YottaDB heap space, each of of size
+  :code:`bufSiz`; and
+- an array of :code:`numSubs` :code:`C.ydb_buffer_t` structures, also
+  in YottaDB heap space, each referencing an allocated buffer.
+
+Set:
+
+- In each :code:`C.ydb_buffer_t` structure:
+
+  - :code:`len_alloc` to :code:`bufSiz`; and
+  - :code:`len_used` to zero.
+  
+- In the :code:`BufferTArray` structure:
+
+  - :code:`cbuftary` to reference the beginning of the :code:`C.ydb_buffer_t` array;
+  - :code:`elemsAlloc` field to :code:`numSubs`; and
+  - :code:`elemsUsed` is to zero.
+
+Go BufferTArray Dump()
+----------------------
+
+.. code-block:: go
+
+    Dump()
+
+For debugging purposes, dump on stdout:
+
+- :code:`cbuftary` as a hexadecimal address;
+- :code:`elemsAlloc` and :code:`elemsUsed` as integers;
+- for each element of the smaller of :code:`elemsAlloc` and
+  :code:`elemsUsed` elements of the :code:`C.ydb_buffer_t` array
+  referenced by :code:`cbuftary`:
+
+  - :code:`buf_addr` as a hexadecimal address, and
+  - :code:`len_alloc` and :code:`len_used` as integers; and
+  - the smaller of :code:`len_used` and :code:`len_alloc` bytes at the
+    address :code:`buf_addr`, in `zwrite format`_.
+
+Go BufferTArray Free()
+----------------------
+
+.. code-block:: go
+
+    Free()
+
+The inverse of the :code:`Alloc()` method: release the :code:`numSubs`
+buffers and the :code:`C.ydb_buffer_t` array. Set :code:`cbuftary` to
+:code:`nil`.
```

Comment:
Will elemsAlloc and elemsUsed also be set to 0?

Reply:
Technically not necessary, because if cbuftary is nil, elemsAlloc and elemsUsed are not meaningful. But I added zeroing them to the specification since that is good defensive programming.